sdk/src/androidTest/java/com/bugsnag/android/ThreadStateTest.kt

```diff
@@ -0,0 +1,160 @@
+
+        verifyCurrentThreadStructure(json, currentThread.id) {
+            assertTrue(it.getJSONArray("stacktrace").length() > 0)
+        }
+    }
+
+    /**
+     * Verifies that a handled error uses [Thread] for the reporting thread stacktrace
+     */
+    @Test
+    fun testHandledStacktrace() {
+        val currentThread = Thread.currentThread()
+        val allStackTraces = Thread.getAllStackTraces()
+        val state = ThreadState(configuration, currentThread, allStackTraces, null)
+        val json = streamableToJsonArray(state)
+
+        // find the stack trace for the current thread that was passed as a parameter
+        val expectedTrace = allStackTraces.filter {
+            it.key.id == currentThread.id
+        }.map { it.value }.first()
+
+        verifyCurrentThreadStructure(json, currentThread.id) {
+
+            // the thread id + name should always be used
+            assertEquals(currentThread.name, it.getString("name"))
+            assertEquals(currentThread.id, it.getLong("id"))
+
+            // stacktrace should come from the thread (check same length and line numbers)
+            val serialisedTrace = it.getJSONArray("stacktrace")
+            assertEquals(expectedTrace.size, serialisedTrace.length())
+
+            expectedTrace.forEachIndexed { index, element ->
+                val jsonObject = serialisedTrace.getJSONObject(index)
+                assertEquals(element.lineNumber, jsonObject.getInt("lineNumber"))
+            }
+        }
+    }
+
+    /**
+     * * Verifies that an unhandled error uses [Exception] for the reporting thread stacktrace
+     */
+    @Test
+    fun testUnhandledStacktrace() {
+        val currentThread = Thread.currentThread()
+        val allStackTraces = Thread.getAllStackTraces()
+        val exc: Throwable = RuntimeException("Whoops")
+        val expectedTrace = exc.stackTrace
+
+        val state = ThreadState(configuration, currentThread, allStackTraces, exc)
+        val json = streamableToJsonArray(state)
+
+        verifyCurrentThreadStructure(json, currentThread.id) {
+
+            // the thread id + name should always be used
+            assertEquals(currentThread.name, it.getString("name"))
+            assertEquals(currentThread.id, it.getLong("id"))
+
+            // stacktrace should come from the exception (check different length)
+            val serialisedTrace = it.getJSONArray("stacktrace")
+            assertEquals(expectedTrace.size, serialisedTrace.length())
+
+            expectedTrace.forEachIndexed { index, element ->
+                val jsonObject = serialisedTrace.getJSONObject(index)
+                assertEquals(element.lineNumber, jsonObject.getInt("lineNumber"))
+            }
+        }
+    }
+
+    private fun verifyCurrentThreadStructure(json: JSONArray,
+                                             currentThreadId: Long,
+                                             action: ((thread: JSONObject) -> Unit)? = null) {
```

Comment:
Allows passing a function that will be invoked when the errorReportingThread json is found. This means assertions can be run with less boilerplate finding the necessary thread